areader/reader.go

```diff
@@ -16,313 +16,418 @@ package areader
+			return nil, errors.Errorf("reader: invalid header element: %v", hdr.Name)
 		}
-		// if not just register worker for given update type
-		p, err := ar.ParseManager.GetRegistered(update.Type)
+		fallthrough
+
+	case "header.tar.gz":
+		hSum, err := ar.readHeader(tReader)
 		if err != nil {
-			// if there is no registered one; check if we can use generic
-			p = ar.ParseManager.GetGeneric(update.Type)
-			if p == nil {
-				return nil, errors.Wrapf(err,
-					"reader: can not find parser for update type: [%v]", update.Type)
+			return nil, err
+		}
+		if hSum != nil {
+			// verify checksums of header
+			hc, err := manifest.Get("header.tar.gz")
+
+			if err != nil {
+				return nil, err
+			}
+			if bytes.Compare(hc, hSum) != 0 {
+				return nil, errors.Errorf("reader: invalid header checksum [%s]:[%s]",
+					hc, hSum)
 			}
 		}
-		ar.ParseManager.PushWorker(p, fmt.Sprintf("%04d", cnt))
+	default:
+		return nil, errors.Errorf("reader: found unexpected file in artifact: %v",
+			hdr.FileInfo().Name())
 	}
-	return ar.ParseManager.GetWorkers(), nil
+	return manifest, nil
 }
 
-func (ar *Reader) ReadInfo() (*metadata.Info, error) {
-	info := new(metadata.Info)
-	err := ar.readNext(info, "version")
+func (ar *Reader) ReadArtifact() error {
+	// each artifact is tar archive
+	if ar.r == nil {
+		return errors.New("reader: read artifact called on invalid stream")
+	}
+	tReader := tar.NewReader(ar.r)
+
+	// first file inside the artifact MUST be version
+	ver, vsum, err := readVersion(tReader)
 	if err != nil {
-		return nil, err
+		return errors.Wrapf(err, "reader: can not read version file")
 	}
-	return info, nil
-}
+	ar.info = ver
+
+	var s *artifact.ChecksumStore
 
-func getUpdateFromHdr(hdr string) string {
-	r := strings.Split(hdr, string(os.PathSeparator))
-	if len(r) < 2 {
-		return ""
+	switch ver.Version {
+	case 1:
+		if err = ar.readHeaderV1(tReader); err != nil {
+			return err
+		}
+	case 2:
+		s, err = ar.readHeaderV2(tReader, vsum)
+		if err != nil {
+			return err
+		}
+	default:
+		return errors.Errorf("reader: unsupported version: %d", ver.Version)
 	}
-	return r[1]
+	return ar.readData(tReader, s)
 }
 
-func (ar *Reader) ReadNextHeader() (parser.Parser, error) {
+func (ar *Reader) GetCompatibleDevices() []string {
+	return ar.hInfo.CompatibleDevices
+}
 
-	var p parser.Parser
+func (ar *Reader) GetArtifactName() string {
+	return ar.hInfo.ArtifactName
+}
 
-	for {
+func (ar *Reader) GetInfo() artifact.Info {
+	return *ar.info
+}
 
-		var hdr *tar.Header
-		hdr, err := getNext(ar.hReader)
-		if err == io.EOF {
-			errClose := ar.Close()
-			if errClose != nil {
-				return nil, errors.Wrapf(errClose, "reader: error closing header reader")
+func (ar *Reader) setInstallers(upd []artifact.UpdateType) error {
+	for i, update := range upd {
+		// firsrt check if we have worker for given update
+		if w, ok := ar.installers[i]; ok {
+			if w.GetType() == update.Type || w.GetType() == "generic" {
```

Comment:
Fixed! This was leftover form the previous implementation, when generic handler has specific update type called `generic`. This is not true any more.